modules/transactionpool/valid.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	return
+}
+
+// validUnconfirmedFileContractTerminations checks that all file contract
+// terminations are valid in the context of the unconfirmed consensus set.
+// There is an additional check for the validity of the unlock conditions and
+// the validity of the termination payouts.
+func (tp *TransactionPool) validUnconfirmedFileContractTerminations(t consensus.Transaction) (err error) {
+	for _, fct := range t.FileContractTerminations {
+		// Check that the file contract has not already been terminated.
+		_, exists := tp.fileContractTerminations[fct.ParentID]
+		if exists {
+			return errors.New("termination given for an already terminated file contract")
+		}
+
+		// Check for the corresponding FileContract in both the confirmed and
+		// unconfirmed set.
+		fc, exists := tp.state.FileContract(fct.ParentID)
+		if !exists {
+			fc, exists = tp.fileContracts[fct.ParentID]
+			if !exists {
+				return errors.New("termination given for unrecognized file contract")
+			}
+		}
+
+		// Check that the unlock conditions are resonable.
+		err = tp.state.ValidUnlockConditions(fct.TerminationConditions, fc.TerminationHash)
+		if err != nil {
+			return
+		}
+
+		// Check that the payouts in the termination add up to the payout of
+		// the contract.
+		var payoutSum consensus.Currency
+		for _, payout := range fct.Payouts {
+			payoutSum = payoutSum.Add(payout.Value)
+		}
+		if payoutSum.Cmp(fc.Payout) != 0 {
+			return errors.New("contract termination has incorrect payouts")
+		}
+	}
+
+	return
+}
+
+// validUnconfirmedSiafunds checks that all siafund inputs are valid in the
+// context of the unconfirmed consensus set and that the value of the siafund
+// inputs matches the value of the siafund outputs. There is also a check on
+// the unlock conditions.
+func (tp *TransactionPool) validUnconfirmedSiafunds(t consensus.Transaction) (err error) {
+	var inputSum consensus.Currency
+	for _, sfi := range t.SiafundInputs {
+		// Check that the siafund output has not already been spent.
+		_, exists := tp.usedSiafundOutputs[sfi.ParentID]
+		if exists {
+			return errors.New("siafund output has already been spent")
+		}
+
+		// Check that the siafund output being spent exists.
+		sfo, exists := tp.state.SiafundOutput(sfi.ParentID)
+		if !exists {
+			sfo, exists = tp.siafundOutputs[sfi.ParentID]
+			if !exists {
+				return errors.New("transaction spends unrecognized siafund output")
+			}
+		}
+
+		// Check that the unlock conditions are reasonable.
+		err = tp.state.ValidUnlockConditions(sfi.UnlockConditions, sfo.UnlockHash)
+		if err != nil {
+			return
+		}
+
+		// Add this input's value to the inputSum.
+		inputSum = inputSum.Add(sfo.Value)
+	}
+
+	// Go through the outputs, making sure each is valid and summing the number
+	// of funds allocated.
+	var outputSum consensus.Currency
+	for _, sfo := range t.SiafundOutputs {
+		if sfo.ClaimStart.Cmp(consensus.ZeroCurrency) != 0 {
+			return errors.New("invalid siafund output presented")
+		}
```

Comment:
Ideally we wouldn't, especially because there are multiple ways to implement something like ClaimStart